### PR TITLE
LibWeb: Update session history when History entry is pushed/replaced 

### PR DESCRIPTION
**Pull request by kalenikaliaksandr**
- Update `HTML::History` to have index and length like latest specification says
- Make `shared_history_push_replace_state()` to actually add/replace session history entry
- Update URL in UI when `pushState()` or `replaceState()` happens
- Add test for `pushState()` :)

Fixes https://github.com/SerenityOS/serenity/issues/21185

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/History.h [hunk 1]
```diff
@@ -25,7 +25,10 @@ class History final : public Bindings::PlatformObject {
     WebIDL::ExceptionOr<void> go(long delta);
     WebIDL::ExceptionOr<void> back();
     WebIDL::ExceptionOr<void> forward();
-    WebIDL::ExceptionOr<u64> length() const;
+    WebIDL::ExceptionOr<size_t> length() const;
+
+    size_t m_index { 0 };
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/History.h:31]: can we go for u64 or u32 here explicitly? there was a great purge of size_t from the code some time ago, be a shame to keep adding more uses.

File: Userland/Libraries/LibWeb/DOM/Document.cpp [hunk 2]
```diff
@@ -3489,4 +3489,37 @@ Painting::ViewportPaintable* Document::paintable()
...
+
+    // 5. If documentsEntryChanged is true, then:
+    if (documents_entry_changed) {
+        // FIXME: Implement this.
+    }
+
+    // 6. If documentIsNew is true, then:
+    if (document_is_new) {
+        // FIXME: 1. Try to scroll to the fragment for document.
+        // FIXME: 2. At this point scripts may run for the newly-created document document.
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/DOM/Document.cpp:3515]: Was there some other spec concept here that we wait for in the parser in  order to run scripts? Or is that all still mega ad-hoc

Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/DOM/Document.cpp:3515]: yes it seems so https://html.spec.whatwg.org/multipage/document-lifecycle.html#read-html:scripts-may-run-for-the-newly-created-document
let's take care of it in a separate PR? :)

File: Userland/Libraries/LibWeb/HTML/History.h (hunk 1 above)
Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/HTML/History.h]: sure, fixed!

File: Userland/Libraries/LibWeb/DOM/Document.cpp (hunk 2 above)
Comment (ADKaster) [reply Userland/Libraries/LibWeb/DOM/Document.cpp:3515]: I was thinking of step 5 here in the end:
<https://html.spec.whatwg.org/multipage/parsing.html#the-end>

But it seems like that new concept is just kinda... shoehorned in? Not sure how it fits into "is ready to be parser executed" and such. So yeah, separate PR :)

**Review by ADKaster — APPROVED** [review]

**Pull request merged**